GUACAMOLE-547: Add support for SSH NONE authentication method

necouchman opened the pull request:
I'm not 100% certain that this is all that's needed, because I don't have a SSH server to test this against, but, according to the libssh2 documentation, this may be all that's required to allow the session to continue when there's no authentication required.

https://libssh2.org/libssh2_userauth_list.html

mike-jumper commented on src/common-ssh/ssh.c:
```diff
@@ -321,6 +321,10 @@ static int guac_common_ssh_authenticate(guac_common_ssh_session* common_session)
             "Supported authentication methods: %s", user_authlist);
 
+    /* If auth list is NULL, then authentication has succeeded with NONE */
+    if (user_authlist == NULL)
```
If `user_authlist` is `NULL`, what happens to that `guac_client_log()` call above that tries to print the value?

necouchman replied:
Good point.  Perhaps that NULL check should be located elsewhere...

necouchman replied:
Relocated and logged the NONE success.

mike-jumper commented:
I'll try to spin up a server and test.

necouchman commented:
> I'll try to spin up a server and test.

If you're able to, can you provide documentation on how to do this?  I spent some time on the Google machine trying to find instructions for it, and could never figure out how to configure OpenSSH to support NONE authentication method.

mike-jumper commented:
OK - I think I figured it out:

1. Prepare a server that you're OK with having essentially no protection.
2. Edit `/etc/ssh/sshd_config` such that the following are set to `yes`:

       PasswordAuthentication yes
       PermitEmptyPasswords yes

3. Create an arbitrary user and use `passwd` to delete their password (making the account passwordless):

       $ sudo useradd test
       $ sudo passwd -d test
       Removing password for user test.
       passwd: Success
       $

Authentication for that specific user is then disabled for SSH and you should be able to log in as that user with any SSH client that supports the "NONE" method.

I tried the above on a fresh and isolated CentOS 7 machine and can reproduce the same behavior described on [GUACAMOLE-547](https://issues.apache.org/jira/browse/GUACAMOLE-547), and can confirm that your changes here allow authentication to succeed.

... **BUT:** With these changes in place, Guacamole still prompts the user for a password, unlike the OpenSSH client which just allows the session to begin. Things probably need further rearranging for Guacamole's behavior to be truly 100% correct for SSH NONE.

mike-jumper commented:
If it helps, I can open up the test SSH server for you to test against. I'd need to know the IP address or subnet you'll be testing from, though.

necouchman commented:
> With these changes in place, Guacamole still prompts the user for a password, unlike the OpenSSH client which just allows the session to begin. Things probably need further rearranging for Guacamole's behavior to be truly 100% correct for SSH NONE.

Okay, I'll spend a little more time trying to rearrange things to see what I can figure out.  I'm guessing the `libssh2_userauth_list()` call just needs to be moved up some (or the call for the password needs to be moved down), so I'll try to rearrange things.

> If it helps, I can open up the test SSH server for you to test against. I'd need to know the IP address or subnet you'll be testing from, though.

Appreciate the offer, I'll let you know if I need it.  I think with your instructions above I should be able to set something up in my test environment that I can use to reproduce this.

mike-jumper commented:
> Okay, I'll spend a little more time trying to rearrange things to see what I can figure out. I'm guessing the `libssh2_userauth_list()` call just needs to be moved up some (or the call for the password needs to be moved down), so I'll try to rearrange things.

Since the password prompt is currently started by the SSH protocol implementation (`src/protocols/ssh/`) and not by the common SSH client (`src/common-ssh/`), I think we may need to provide some sort of callback to allow `guac_common_ssh_authenticate()` to request additional credentials which the caller can then satisfy.

Some similar means for the caller to respond to additional credential requests from the SSH server would also eventually be necessary for [GUACAMOLE-141](https://issues.apache.org/jira/browse/GUACAMOLE-141) ("Complete support for keyboard interactive authentication").

necouchman commented:
> Since the password prompt is currently started by the SSH protocol implementation (src/protocols/ssh/) and not by the common SSH client (src/common-ssh/), I think we may need to provide some sort of callback to allow guac_common_ssh_authenticate() to request additional credentials which the caller can then satisfy.

Okay, sounds like a good challenge :-).  I'll see what I can figure out.

necouchman commented:
Followed your instructions for configuring a user that results in NONE authentication, so will try to figure out the best way to organize things to avoid the password prompt.  Presumably it would also be nice to do this in a way that the SFTP support would work, as well, both for SSH connections and for VNC or RDP using SFTP for file transfer.

necouchman commented:
Well, I'm trying to make this work, but having trouble figuring out the best place to put the callback function.  I was trying to re-use the existing `guac_terminal_prompt` function and put it in the `guac_common_ssh_session` struct, but then I have to have the terminal, which means retrieving the `guac_ssh_client` struct from within the common SSH code, which seems like something that shouldn't be done.  I'll keep playing with it, but, in the meantime, I'd be happy to take any advice or direction.

necouchman commented:
Okay, this now works with the call-back function, skipping the password prompt for the NONE authentication method but prompting for a password if NONE and public key methods fail, and then using that password for password or keyboard-interactive.  It also sets the credential for use with the SFTP connection, and should correctly skip the SFTP connection on other protocols if a password is not provided (though I believe this will still kill the overall connection).

I'm sure it needs some fixing up, but hopefully it's going in the right direction?

necouchman commented:
@mike-jumper What else needs to be fixed, here?

mike-jumper commented on src/protocols/ssh/ssh.c:
```diff
@@ -130,26 +130,36 @@ static guac_common_ssh_user* guac_ssh_get_user(guac_client* client) {
 }
 
+/**
+ * A call-back function used to gather additional credentials from a client
```
This looks like documentation intended for the `guac_ssh_credential_handler` type. It makes sense from a generic perspective, but the information provided is only sufficient for a developer trying to implement such a callback, not for a developer potentially wanting to use this existing implementation.

It should be clear from the documentation of the function why this particular implementation is appropriate from the perspective of a developer that may want to use this. It prompts the user for credentials using the terminal emulator associated with the SSH connection via `guac_client`? Probably good to note.

necouchman replied:
Reworded this a bit and added the notes as suggested.

mike-jumper commented on src/protocols/ssh/ssh.c:
```diff
@@ -130,26 +130,36 @@ static guac_common_ssh_user* guac_ssh_get_user(guac_client* client) {
+ * @return 
+ *     The string of credentials gathered from the user.
+ */
+char* guac_ssh_get_credential(guac_client *client, char* credName) {
```
See above - the C convention for variables is `lowercase_with_underscores`.

necouchman replied:
Changed, here, too.

mike-jumper commented on src/common-ssh/ssh.c:
```diff
@@ -416,7 +427,7 @@ static int guac_common_ssh_authenticate(guac_common_ssh_session* common_session)
 guac_common_ssh_session* guac_common_ssh_create_session(guac_client* client,
         const char* hostname, const char* port, guac_common_ssh_user* user, int keepalive,
-        const char* host_key) {
+        const char* host_key, guac_ssh_credential_handler* credential_handler) {
```
This is called `credential_callback` in the prototype. Either is fine with me, but the prototype and function body should be consistent.

necouchman replied:
Went with `credential_handler` across the board.

necouchman commented on src/common-ssh/common-ssh/ssh.h:
```diff
@@ -25,6 +25,11 @@
 #include <libssh2.h>
 
+/**
+ * Handler for retrieving additional credentials.
```
Documented.

necouchman commented on src/common-ssh/common-ssh/ssh.h:
```diff
@@ -25,6 +25,11 @@
+/**
+ * Handler for retrieving additional credentials.
+ */
+typedef char* guac_ssh_credential_handler(guac_client* client, char* credName);
```
good_now

necouchman commented on src/common-ssh/common-ssh/ssh.h:
```diff
@@ -99,7 +109,7 @@ void guac_common_ssh_uninit();
 guac_common_ssh_session* guac_common_ssh_create_session(guac_client* client,
         const char* hostname, const char* port, guac_common_ssh_user* user, int keepalive,
-        const char* host_key);
+        const char* host_key, guac_ssh_credential_handler* credential_callback);
```
Documented the new one, plus all of the missing ones.

necouchman commented on src/common-ssh/ssh.c:
```diff
@@ -352,14 +359,18 @@ static int guac_common_ssh_authenticate(guac_common_ssh_session* common_session)
 
     }
 
+    /* Down to username + password authentication. */
```
Down with the rewording.

necouchman commented on src/common-ssh/ssh.c:
```diff
@@ -304,23 +304,30 @@ static int guac_common_ssh_authenticate(guac_common_ssh_session* common_session)
     guac_client_log(client, GUAC_LOG_DEBUG,
             "Supported authentication methods: %s", user_authlist);
 
+
```
Ummm....because...

Removed.

necouchman commented:
@mike-jumper Okay, fixed up the stuff you requested - let me know if anything else needs work.  Also rebased on top of current `staging/1.1.0` branch.

mike-jumper commented on src/common-ssh/common-ssh/ssh.h:
```diff
@@ -25,6 +25,20 @@
+ * @param cred_name
+ *     The name of the credential being requested, which will be shared
+ *     with the client in order to generate a meaningful prompt.
+ * 
```
Missing docs for the return value.

If the returned string will eventually be freed with a call to `free()` and is thus required to be dynamically allocated, that should be documented as well.

necouchman replied:
Added that documentation.

mike-jumper replied:
Looks better, but:

> ... which should be a dynamically-allocated such that it can be freed as required.

More than that, the returned value _must_ be dynamically-allocated, as the caller will eventually attempt to free it with a call to `free()`. For example:

https://github.com/apache/guacamole-server/blob/5e2ddb890a90e840a419f4878381b0c511cc6f25/src/terminal/terminal/terminal.h#L664-L670

and

https://github.com/apache/guacamole-server/blob/6dad6cd91933a19bc18095854e93354b6764d763/src/protocols/rdp/tests/fs/normalize_path.c#L174-L179

While we're here: is there a meaning for returning `NULL`? If not, that's fine, and the documentation covers what it needs to cover. Well-behaved implementations won't return `NULL` unless it's actually documented as a legal return value. However, if an implementation returning `NULL` is intended and has a specific meaning, that should be documented here, too.

necouchman replied:
Okay, reworked.

mike-jumper commented on src/common-ssh/common-ssh/ssh.h:
```diff
@@ -92,14 +114,31 @@ void guac_common_ssh_uninit();
+ * 
+ * @param credential_handler
+ *     The handler function for retrieving additional credentials from the user
+ *     as required by the SSH server.
```
Probably worth adding "or NULL if ...", as that is an intended use.

necouchman replied:
Added that bit in.
